### PR TITLE
SFT-83 As VisComm, the OnboardMediaPlay has clean architecture and code

### PR DESCRIPTION
**Pull request by mglyngberg**
Fixes SFT-83 As VisComm, the OnboardMediaPlay has clean architecture and code
Fixes SFT-80 As VisComm, I want to replace dirent with QDir in onBoardMediaPlayer
Fixes SFT-83 Creater I_SongPlayerPresenter
Fixes SFT-84 Create I_SongUrlSource
Fixes SFT-85 Create I_SongLoader

The threaded brightest color functionality was not included in this, it can be added back in for a future story.

### CODE REVIEW

**Review by ThomasVy — APPROVED** [review]
We did good

File: src/BusinessLayer/SongController/SongController.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+void SongController::playNext()
+{
+    if (songUrls_.size() == 0)
+    {
+        return;
+    }
+
+    //Add the current song to previously played songs
```
Comment (navjotbrar) [inline src/BusinessLayer/SongController/SongController.cpp]: Remove comment

File: src/ViewLayer/SongPlayerView/SongControl/SongControlView.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+SongControlView::SongControlView
+(
+    I_SongControlPresenter& songControlPresenter,
+    I_SongControlUi& songControlUi
+):
+    songControlPresenter_(songControlPresenter),
+    songControlUi_(songControlUi),
+    songState_(songControlPresenter_.songPlayerState())
+{
+    //Connecting buttons to data
```
Comment (navjotbrar) [inline src/ViewLayer/SongPlayerView/SongControl/SongControlView.cpp]: Remove Comment

File: src/ViewLayer/SongPlayerView/SongControl/SongControlView.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    connect(&songControlUi_.playButton(), SIGNAL(clicked()), this, SLOT(playButtonClicked()));
+    connect(&songControlUi_.nextSongButton(), SIGNAL(clicked()), this, SLOT(nextSongButtonClicked()));
+    connect(&songControlUi_.prevSongButton(),  SIGNAL(clicked()), this, SLOT(prevSongButtonClicked()));
+    connect(&songControlUi_.shuffleButton(),  SIGNAL(clicked()), this, SLOT(shuffleButtonClicked()));
+    connect(&songControlUi_.loopButton(),  SIGNAL(clicked()), this, SLOT(loopButtonClicked()));
+    connect(&songControlUi_.volumeControlSlider(), SIGNAL(valueChanged(int)), this, SLOT(volumeSliderMoved(int)));
+
+    //Connecting states to UI
```
Comment (navjotbrar) [inline src/ViewLayer/SongPlayerView/SongControl/SongControlView.cpp]: Remove Comment

**Review by navjotbrar — CHANGES_REQUESTED** [review]
Great Work, Just need to remove some comments

File: src/BusinessLayer/Presenter/SongControl/SongControlPresenter.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#pragma once
+
+#include "I_SongControlPresenter.h"
+
+#include <QObject>
+
+class I_SongPlayerState;
+class I_SongController;
+
+class SongControlPresenter : public QObject, public I_SongControlPresenter
```
Comment (bill-luu) [inline src/BusinessLayer/Presenter/SongControl/SongControlPresenter.h]: Do you need to inherit `QObject` here? Might not need the Q_OBJECT macro either.

File: src/BusinessLayer/SongController/SongController.cpp (hunk 1 above)
Comment (ThomasVy) [reply src/BusinessLayer/SongController/SongController.cpp]: I don't completely agree, but I'll remove the comments.

File: src/ViewLayer/SongPlayerView/SongControl/SongControlView.cpp (hunk 2 above)
Comment (ThomasVy) [reply src/ViewLayer/SongPlayerView/SongControl/SongControlView.cpp]: I don't completely agree with the removal of comments, but I will remove them.

File: src/BusinessLayer/Presenter/SongControl/SongControlPresenter.cpp [hunk 5]
```diff
@@ -0,0 +1,29 @@
+#include "SongControlPresenter.h"
+#include "I_SongPlayerState.h"
+#include "I_SongController.h"
+
+SongControlPresenter::SongControlPresenter(I_SongController& songController,
+        I_SongPlayerState& songState)
+    : songController_(songController),
+      songState_(songState)
```
Comment (bill-luu) [inline src/BusinessLayer/Presenter/SongControl/SongControlPresenter.cpp]: Line up the commas with the colon
```
    : songController_(songController)
    , songState_(songState)
```

File: src/BusinessLayer/SongController/SongController.cpp [hunk 6]
```diff
@@ -0,0 +1,96 @@
+#include "SongController.h"
+#include "I_SongPlayerState.h"
+#include "I_SongPlayer.h"
+#include "I_SongUrlSource.h"
+
+#include <QMediaContent>
+#include <QRandomGenerator>
+#include <QStack>
+
+SongController::SongController(I_SongPlayer& songPlayer, I_SongPlayerState& songPlayerState,
```
Comment (bill-luu) [inline src/BusinessLayer/SongController/SongController.cpp]:
```
SongController::SongController(I_SongPlayer& songPlayer,
                               I_SongPlayerState& songPlayerState,
                               I_SongUrlSource& songUrlSource)
```

File: src/BusinessLayer/SongPlayer/I_SongPlayer.h [hunk 7]
```diff
@@ -1,12 +1,19 @@
 #pragma once
 
-#include <QWidget>
+#include <QMediaPlayer>
+#include<QObject>
```
Comment (bill-luu) [inline src/BusinessLayer/SongPlayer/I_SongPlayer.h]: `#include <QObject>`

File: src/BusinessLayer/SongPlayerState/SongControlEntity.h.autosave [hunk 8]
```diff
@@ -0,0 +1,35 @@
+#pragma once
```
Comment (bill-luu) [inline src/BusinessLayer/SongPlayerState/SongControlEntity.h.autosave]: Delete this file

File: src/OnboardMediaControl/OnboardMediaControl.pro [hunk 9]
```diff
@@ -1,17 +1,21 @@
 TEMPLATE = app
-LIBS += -L../ViewLayer/.lib -L../BusinessLayer/.lib -lViewLayer -lBusinessLayer
+LIBS += -L../ViewLayer/.lib -L../BusinessLayer/.lib -lViewLayer -lBusinessLayer -L../External/.lib -lExternal
```
Comment (bill-luu) [inline src/OnboardMediaControl/OnboardMediaControl.pro]: For consistency:
`LIBS += -L../ViewLayer/.lib -L../BusinessLayer/.lib -L../External/.lib -lViewLayer -lBusinessLayer -lExternal`

File: src/ViewLayer/SongPlayerView/SongControl/SongControlView.cpp [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+void SongControlView::volumeSliderMoved(int newVolume)
+{
+    songControlPresenter_.songPlayerState().setVolume(newVolume);
+}
+
+void SongControlView::initializeUi()
+{
+    songControlUi_.playButton().setCheckable(true);
+    songControlUi_.loopButton().setCheckable(true);
+    songControlUi_.shuffleButton().setCheckable(true);
```
Comment (bill-luu) [inline src/ViewLayer/SongPlayerView/SongControl/SongControlView.cpp]: Can't you do these in the Ui file?

**Review by bill-luu — CHANGES_REQUESTED** [review]
Global things:

1. Includes should be in general format of:
```
#include <QStuff>
#include <Librarys.

#include "OurFiles.h"
```

2. Includes should be in alphabetical order :)

File: src/BusinessLayer/Presenter/SongControl/SongControlPresenter.h (hunk 4 above)
Comment (ThomasVy) [reply src/BusinessLayer/Presenter/SongControl/SongControlPresenter.h]: You're right, I don't think we need QObject.  I'll remove it.

File: src/BusinessLayer/Presenter/SongControl/SongControlPresenter.cpp (hunk 5 above)
Comment (ThomasVy) [reply src/BusinessLayer/Presenter/SongControl/SongControlPresenter.cpp]: Changed!

File: src/BusinessLayer/SongController/SongController.cpp (hunk 6 above)
Comment (ThomasVy) [reply src/BusinessLayer/SongController/SongController.cpp]: Changed!

File: src/BusinessLayer/SongPlayer/I_SongPlayer.h (hunk 7 above)
Comment (ThomasVy) [reply src/BusinessLayer/SongPlayer/I_SongPlayer.h]: I have no idea how that happened, but changed!

File: src/BusinessLayer/SongPlayerState/SongControlEntity.h.autosave (hunk 8 above)
Comment (ThomasVy) [reply src/BusinessLayer/SongPlayerState/SongControlEntity.h.autosave]: Okay :(

File: src/OnboardMediaControl/OnboardMediaControl.pro (hunk 9 above)
Comment (ThomasVy) [reply src/OnboardMediaControl/OnboardMediaControl.pro]: Okay

File: src/ViewLayer/SongPlayerView/SongControl/SongControlView.cpp (hunk 10 above)
Comment (ThomasVy) [reply src/ViewLayer/SongPlayerView/SongControl/SongControlView.cpp]: You're correct, sir. I set checkable in the ui file and removed initializeUi method

**Review by navjotbrar — APPROVED** [review]
Great Work!

File: README.md [hunk 11]
```diff
@@ -0,0 +1,5 @@
+# Epsilon Onboard Media Player
```
Comment (bill-luu) [inline README.md:1]: There should also be some more instruction regarding where the songs should go!

**Review by bill-luu — CHANGES_REQUESTED** [review]
Round 2

File: README.md [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+
+This program is the onboard media control for the University of Calgary’s Solar Car Schulich Elysia. This program is a music player which runs on the car's secondary Raspberry Pi on the center display.
```
Comment (bill-luu) [inline README.md]: Yo is it me or:
```
...University of Calgary’s...
```
uses a different apostrophe than
```
...which runs on the car's...
```

**Review by bill-luu — APPROVED** [review]
Great job! It was a big effort, but the code looks much cleaner overall. Just clean up the commit history a little bit, but otherwise that's an approval from me. 
image

**Pull request merged**